### ISSUE TITLE
Main Trainer

### ISSUE THREAD

**Issue by mousyball** [mousyball/pytorch-trainer#3]
List the dev note here.

**Comment by JanLin0817** [comment]
```python
class Hook:
    def before_train_epoch(self, runner):
        self.before_epoch(runner)

    def before_val_epoch(self, runner):
        self.before_epoch(runner)
```
train/val have **different** Hook  behavior, child class implement `before_train_epoch` and `before_train_epoch` separately

**Comment by JanLin0817** [comment]
Variables in the trainer
* `epoch` : epoch
* `max_iter` :  iteration maximum
* `max_epoch` :  epoch maximum 
* `iter` : batch number accumulate through epochs
* `inner_iter` :  batch number accumulate within epochs

**Comment by JanLin0817** [comment]
remember optimizer.step() and scheduler.step() have priority
scheduler.step() should come after optimizer.step()

**Comment by JanLin0817** [comment]
## Scope between hook and trainer
### trainer
* loss_meter
* text logger (logging)

### LoggerHook
* update/clear trainer meter
* put text logger in hook or trainer

**Comment by JanLin0817** [comment]
## Distribution train
* sync_buffer: Whether to synchronize buffers between GPUs

**Comment by JanLin0817** [comment]
## management

* According to [mmcv-registry](https://github.com/open-mmlab/mmcv/blob/080474b9a6387a9c617e69ffcd4c091ed3bd4d4b/docs/registry.md), we should provide an api or document to list all register function in all namespace
* [Detectron2 - register](https://github.com/facebookresearch/detectron2/blob/ac96db61151f75758f925124d7f0656cc545c977/docs/modules/modeling.rst)

**Comment by JanLin0817** [comment]
## TODO list
* [x] optimizer.step() scheduler.step() have priority
  * scheduler.step() should come after optimizer.step()
* [x] handle 'module' from loaded state_dict
* [ ] sync_buffer: Whether to synchronize buffers between GPUs
* [ ] Make this [doc](https://github.com/mousyball/pytorch-trainer/blob/develop/docs/priority.md) clear
* [ ] Rename function which end with "_batch"
* [ ] Extract test model to a single file
* [ ] single/multi GPU trainer
* [ ] distribution trainer
